### ISSUE TITLE
<feature> CMDB layered file system

### ISSUE THREAD

**Issue by ml019** [codeontap/gen3-freemarker-wrapper#11]
### Background

I've been mulling over how we move our CMDB processing forward. A few things I wanted to achieve
1. Increased speed of processing - avoid all the bash processing and assembly of json files
2. Simple cmdb structure - avoid adding in extra directory levels to handle differing CMDB granularity
3. Read and write - want to be able to not only read CMDBs, but potentially add documents to them as well.

### CMDB Layered File System

I'm proposing we introduce the idea of a logical CMDB file system, much like the unix file system, where each repo is "mounted" at a particular base path. Further, multiple repos can be mounted at the same path (hence the idea of layers, borrowing from the docker layered image). So I can have separate config and infrastructure repos but mount them at the same point in the tree. Thus whether it is one or more than one repo, the logical tree is the same.

For speed, I'm proposing that the Freemarker wrapper implement the CMDB file system. This will permit directory scanning and file reading/writing as part of normal template processing by exposing CMDB access functions via Freemarker. This will also permit reading AND writing to the CMDB, should we want to use the Freemarker engine to generate new files and save these to the CMDB.

While initially we will support physical file system backed CMDBs, the use of a file system abstraction means adding other sources (e.g. database backed cmdbs) can easily be accommodated in the future. 

### .cmdb extensions

The .cmdb file will be extended to indicate any CMDBs that could be layered under the existing CMDB. To avoid overlaps, file lookups will be sought in the base CMDB first, then in the layered CMDBs in the order they are specified. (Overlaps can also be avoided by selection of non-overlapping base paths). Each layer entry will indicate the CMDB required by its name, and the base path in the CMDB file system at which it should be mounted. Paths can be absolute or relative. Relative paths are appended to the base path for the CMDB containing the layer. CMDBs will thus be nestable to arbitrary levels.

```
{
  "Layers" : [
    {
      "Name" : "appointments",
      "BasePath" : "/products/appointments"
    }
  ]
}
```

### Freemarker switches

Freemarker with have new command line switches to support assembling the CMDB file system.

The **"-g"** switch will support specification of the mapping of CMDB names to physical paths on the file system. Multiple -g switches instances will be supported. It will accept one or more values with two formats will be accepted, 

- name=path
- path, where path cannot contain an "=" character

The first form names and locates a single CMDB. 

The second form identifies a directory whose subtree is scanned for .cmdb files, with the containing directory being treated as a CMDB whose name is that of the containing directory. It is expected that a top-level set of CMDB directories will be the most common layout used. 

Note that where CMDB directories are clones of git repos, the CMDB name need not be the same as the repo name, with the management of repo cloning/updating being out of scope as far as Freemarker is concerned.

The **"-c"** switch will define the CMDBs to be processed. Multiple -c switch instances will be supported. It will accept one or more values, each of which is treated as a CMDB name. If the -c switch is not provided, all CMDBs found via the -g switch will be considered available for processing.

The **"-b"** switch will define the base CMDB considered to be at the root of the CMDB structure, so the CMDB with a path of "/". If the switch is not provided, all CMDBs considered for processing will be mounted at "/default".

The three switches allow a variety of arrangements;

- for automation processing, the -g switch could specify the tenancy and product/environment level CMDBs, with the -b switch indicating the tenancy CMDB as the starting point.

- for local processing, the -g switch could point to a local parent directory under which all repos for a tenant would be cloned. The -c switch would then indicate the product to be processed, and the -b default value used to select the tenant CMDB as the starting point. 

### Processing functions

Freemarker will offer several functions for manipulating the CMDB file system.

**getCMDBTree**

```
[#local candidates =
  getCMDBTree(
    path,
    {
      "Regex" : []
      "IgnoreDotDirectories" : true,
      "IgnoreDotFiles" : true,
      "IncludeCMDBInformation" : false
    }
  ) ]
```
where 
- path is the absolute starting point within the CMDB file system to be scanned
- the second parameter is a hash of options
    - Regex is an array of regular expressions to match the filename (default .*)
    - IgnoreDotDirectories controls if directories starting with "." should be ignored (default true)
    - IgnoreDotFiles controls if files starting with "." should be ignored (default true)
    - IncludeCMDBInformation includes cmdb information in the results (default false)

Regex will be used for limiting file extensions or looking for specific elements in file paths e.g. ```.../asFile/...```. The entries of the array are considered to be ORed, mainly to keep the regular expressions simple.  

The Java regex library has been used and searches need to be explicitly anchored if required. If no "^" anchoris provided, then the path must match the beginning of any string before it is considered for regex analysis, and ".*" is inserted before the regex. If a "^" anchor is provided, then the path is ignored. If no "$" anchor is present, the ".*" is postpended to any provided regex.

It should return a list of objects, one for each file found, modelled along the lines of .get_optional_template...
```
{
  "File":
  "Path" : 
  "Filename" :
  "Extension" :
  "Contents" : 
  "ContentsAsJSON" :
  "Include" :
  "CMDB" : { 
    "Name" : <name from -g/-c/-b>,
    "BasePath" :  <absolute path to cmdb in CMDB File System>,
    "ContentsAsJSON": <contents of .cmdb file if present>
    "File" : <full physical path to file>
  }
}
```
File is the absolute path to the file.

Path, Filename and Extension (no dot) reflect the found file details. Filename includes the Extension but not the Path. Path does not include the filename.

Contents is a function that will return the file contents as a string in the same way as the current -v option works. If the contents can be converted to JSON, the ContentsAsJSON attribute will be present permitting direct access to the parsed result.  

If the file is considered to be a Freemarker template (First line = [#ftl]), then there will also be an "Include" attribute present that works the same as its equivalent in .get_optional_template. 

This function will form the basis of processing to assemble the in-memory structure to be processed via component templates.

**getCMDBs**

```
[#local cmdbs =
  getCMDBs(
    {
      "ActiveOnly" : false
    }
   ) ]
```
where 
- the first parameter is a hash of options
    - ActiveOnly limits listed CMDBs to those that are to be considered for processing (default false)

This function will return a list of the cmdbs known to the engine;

```
{
  "Name" : 
  "CMDBPath" :
  "FileSystemPath" :
  "Base" :
  "Active" :
  "ParentCMDB" :
  "ContentsAsJSON":

}
```
Name is the CMDB name.

CMDBPath is the mount point for the cmdb in the CMDB File system.

FileSystemPath is the physical location of the CMDB on disk.

Base is a boolean indicating whether the CMDB is considered the base CMDB. If this flag is true, then CMDBPath will always be "/".

Active is a boolean indicating whether the CMDB is to be considered for processing.

ParentCMDB is the name of the parent CMDB by which the CMDB was mounted. This means it was listed in an entry of the Layers array within the .cmdb file of the parent CMDB.

ContentsAsJSON is the contents of the .cmdb file if present.

**Comment by ml019** [comment]
@kshychko I have been thinking about the use of getFileTree() and how to make it flexible in the face of a variable nesting of cmdbs. As a result, I've added an extra attribute to the CMDB information. The extra parameter is the absolute path to the mount point of the cmdb in the CMDB file system.

My thinking is that if I locate say a tenant.json file, I know that everything about that tenant will be located below the absolute path of the cmdb in the CMDB file system, because that is the way layers work. Thus, I can easily scope a search for products for a particular tenant by using the BasePath of the CMDB for the tenant.json file as the value of the path parameter to the call to getFileTree().

Please let me know when this attribute is available.

**Comment by ml019** [comment]
@kseniya I'd like to get the support for the layered file system we have so far into the master, and also look to use it to do some direct file lookup for stack outputs and settings in the short term. These functions have already been simplified to basic file lookups in the bash side, so should be relatively straightforward to fully perform inside the freemarker engine.

I thus need a version of the layered file system support as follows;
1. includes any fixes that have been made on master
2. has support for Content, Includes and CMDBInformation in the output
2. If the -g switch is absent, then getFileTree() should just return an empty hash and should not throw any errors in regards to absence of other layer related command line options
3. If the base cmdb has no Layers information, create a default Layers structure as follows for each cmdb found to be present (either explicitly via -c/-b options or implicitly via -g option)
```
{
  "Layers" : [
    {
      "Name" : "{cmdb name}",
      "BasePath" : "/default/{cmdb name}"
    }
  ]
}
```
4. For now, ignore any directory paths to cmdbs when processing the -g switch. We will worry about duplicate cmdb names later. (I'm thinking we use a namespace approach for this)

With the above in place, I think we can 
1. Configure the -g switch based on the value of findGen3RootDir
2. don't worry about -c and -b for now and use default behaviour in this case. I will need to tweak constructTree.sh a little bit but aligning it to work with getFileTree() shouldn't be too hard

**Comment by ml019** [comment]
Closing as basic functionality released as version 1.9.20. Separate tickets will be raised for bugs/new features.
